gef.py

```diff
@@ -6760,55 +6749,136 @@ def do_invoke(self, argv):
             info("No Tcache in this version of libc")
             return
 
-        arena = GlibcArena("*{:s}".format(argv[0])) if len(argv) == 1 else get_main_arena()
-
-        if arena is None:
-            err("Invalid Glibc arena")
+        current_thread = gdb.selected_thread()
+        if current_thread is None:
+            err("Couldn't find current thread")
             return
 
-        # Get tcache_perthread_struct for this arena
-        heap_base = HeapBaseFunction.heap_base()
-        if heap_base is None:
-            err("No heap section")
-            return
-        addr = heap_base + 0x10
-
-        gef_print(titlify("Tcachebins for arena {:#x}".format(int(arena))))
-        for i in range(GlibcArena.TCACHE_MAX_BINS):
-            if get_libc_version() < (2, 30):
-                count = ord(read_memory(addr + i, 1))
+        threads = gdb.selected_inferior().threads()
+        if argv:
+            if "all" in argv:
+                tids = [t.num for t in threads]
             else:
-                count = u16(read_memory(addr + 2 * i, 2))
-            chunk = arena.tcachebin(i)
-            chunks = set()
-            m = []
+                tids = self.check_thread_ids(argv)
+        else:
+            tids = [current_thread.num]
 
-            # Only print the entry if there are valid chunks. Don't trust count
-            while True:
-                if chunk is None:
-                    break
+        # As a nicety, we want to display threads in ascending order by gdb number
+        for thread in sorted(threads, key=lambda t: t.num):
+            if thread.num not in tids:
+                continue
 
-                try:
-                    m.append("{:s} {:s} ".format(LEFT_ARROW, str(chunk)))
-                    if chunk.address in chunks:
-                        m.append("{:s} [loop detected]".format(RIGHT_ARROW))
+            thread.switch()
+            try:
+                tcache_addr = gdb.parse_and_eval("(void *) tcache")
+            except gdb.error:
+                # In binaries not linked with pthread, we can't use the tcache
+                # symbol.
+                tcache_addr = self.get_main_thread_tcache()
```

Comment:
instead of putting this logic here, why not make `get_main_thread_tcache` try the parse and eval trick?

Reply:
Really just a question of verbiage - the parse and eval trick gets the tcache address for the current thread, whereas `get_main_thread` will _always_ get the `tcace` from the main thread.

Collapsed the logic into one function and renamed `get_main_thread` in 9849ff90adfa392131c38257a8d78c045ef0e92b